src/mscorlib/shared/System/ReadOnlyMemory.cs

```diff
@@ -0,0 +1,273 @@
+        public ReadOnlyMemory(T[] array, int start, int length)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = array;
+            _index = start;
+            _length = length;
+        }
+        
+        // Constructor for internal use only.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlyMemory(OwnedMemory<T> owner, int index, int length)
+        {
+            if (owner == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.ownedMemory);
+            if (index < 0 || length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = owner;
+            _index = index | (1 << 31); // Before using _index, check if _index < 0, then 'and' it with RemoveOwnedFlagBitMask
+            _length = length;
+        }
+
+        /// <summary>
+        /// Defines an implicit conversion of an array to a <see cref="Memory{T}"/>
+        /// </summary>
+        public static implicit operator ReadOnlyMemory<T>(T[] array) => new ReadOnlyMemory<T>(array);
+        
+        /// <summary>
+        /// Defines an implicit conversion of a <see cref="ArraySegment{T}"/> to a <see cref="Memory{T}"/>
+        /// </summary>
+        public static implicit operator ReadOnlyMemory<T>(ArraySegment<T> arraySegment) => new ReadOnlyMemory<T>(arraySegment.Array, arraySegment.Offset, arraySegment.Count);
+
+        /// <summary>
+        /// Returns an empty <see cref="Memory{T}"/>
+        /// </summary>
+        public static ReadOnlyMemory<T> Empty => Array.Empty<T>();
+
+        /// <summary>
+        /// The number of items in the memory.
+        /// </summary>
+        public int Length => _length;
+
+        /// <summary>
+        /// Returns true if Length is 0.
+        /// </summary>
+        public bool IsEmpty => _length == 0;
+
+        /// <summary>
+        /// Forms a slice out of the given memory, beginning at 'start'.
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index is not in range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlyMemory<T> Slice(int start)
+        {
+            if ((uint)start > (uint)_length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            if (_index < 0)
+                return new ReadOnlyMemory<T>((OwnedMemory<T>)_arrayOrOwnedMemory, (_index & RemoveOwnedFlagBitMask) + start, _length - start);
```

Comment:
:question: Is it possible to wrap this cast in a property that uses the `Unsafe` class to avoid type checks? If so, would this have a performance win that justifies the use of unsafe? /cc @stephentoub

Reply:
Races in user code can make index bit to be out of sync with `_arrayOrOwnedMemory`. Casts are useful to diagnose this situation before it corrupts the system. If the casts are replaced by Unsafe casts, this situation will tend lead to very hard to diagnose crash during GC.

Reply:
Races in user code can make index bit to be out of sync with `_arrayOrOwnedMemory`. Casts are useful to diagnose this situation before it corrupts the system. If the casts are replaced by Unsafe casts, this situation will tend lead to very hard to diagnose crash during GC.